File: 16-3sum-closest/cpp/main.cpp
class Solution {
 public:
  int threeSumClosest(vector<int>& nums, int target) {
    sort(nums.begin(), nums.end());
    int result = nums[0] + nums[1] + nums[2];

    for (int l = 0; l < nums.size(); ++l) {
      int m = l + 1;
      int r = nums.size() - 1;

      while (m < r) {
        int currentSum = nums[l] + nums[m] + nums[r];

        if (currentSum > target) {
          --r;
        } else {
          ++m;
        }

        if (abs(target - currentSum) < abs(target - result)) {
          result = currentSum;
        }
      }
    }

    return result;
  }
};

File: 16-3sum-closest/python/main.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        result = sum(nums[:3])

        for l in range(len(nums)):
            m = l + 1
            r = len(nums) - 1

            while m < r:
                currentSum = nums[l] + nums[m] + nums[r]

                if currentSum > target:
                    r -= 1
                else:
                    m += 1

                if abs(target - currentSum) < abs(target - result):
                    result = currentSum

        return result
